ambry-replication/src/main/java/com.github.ambry.replication/ReplicationEngine.java

```diff
@@ -400,4 +400,60 @@ protected boolean isTokenForRemoteReplicaInfo(RemoteReplicaInfo remoteReplicaInf
         .getReplicaPath()
         .equals(tokenInfo.getReplicaPath());
   }
+
+  /**
+   * Reload replication token if exist for all remote replicas of {@code localReplicaId}.
+   * @param localReplicaId local {@link ReplicaId} for which to reload tokens.
+   * @param remoteReplicaInfos {@link List} of {@link RemoteReplicaInfo} of the {@code localReplicaId}.
+   * @return Number of remote replicas for which reload failed.
+   */
+  protected int reloadReplicationTokenIfExists(ReplicaId localReplicaId, List<RemoteReplicaInfo> remoteReplicaInfos)
+      throws ReplicationException {
+    int tokenReloadFailureCount = 0;
+    List<RemoteReplicaInfo.ReplicaTokenInfo> tokenInfos = persistor.retrieve(localReplicaId.getMountPath());
+    if (tokenInfos.size() != 0) {
+      for (RemoteReplicaInfo remoteReplicaInfo : remoteReplicaInfos) {
+        boolean tokenReloaded = false;
+        for (RemoteReplicaInfo.ReplicaTokenInfo tokenInfo : tokenInfos) {
+          if (isTokenForRemoteReplicaInfo(remoteReplicaInfo, tokenInfo)) {
+            logger.info("Read token for partition {} remote host {} port {} token {}", localReplicaId.getPartitionId(),
+                tokenInfo.getHostname(), tokenInfo.getPort(), tokenInfo.getReplicaToken());
+            tokenReloaded = true;
+            remoteReplicaInfo.initializeTokens(tokenInfo.getReplicaToken());
+            remoteReplicaInfo.setTotalBytesReadFromLocalStore(tokenInfo.getTotalBytesReadFromLocalStore());
+            break;
+          }
+        }
+        if (!tokenReloaded) {
+          // This may happen on clusterMap update: replica removed or added.
+          // Or error on token persist/retrieve.
+          logger.warn("Token not found or reload failed. remoteReplicaInfo: {} tokenInfos: {}", remoteReplicaInfo,
+              tokenInfos);
+          tokenReloadFailureCount++;
+        }
+      }
+    }
+    return tokenReloadFailureCount;
+  }
+
+  /**
+   * Stop replication for the specified {@link PartitionId}.
+   * @param partitionId {@link PartitionId} for which replication should be removed.
+   */
+  protected void stopPartitionReplication(PartitionId partitionId) {
+    PartitionInfo partitionInfo = partitionToPartitionInfo.remove(partitionId);
+    if (partitionInfo == null) {
+      logger.warn("Partition {} not exist when remove from {}. ", partitionId, dataNodeId);
+      return;
+    }
+    removeRemoteReplicaInfoFromReplicaThread(partitionInfo.getRemoteReplicaInfos());
+    if (replicationConfig.replicationPersistTokenOnShutdownOrReplicaRemove) {
+      try {
+        persistor.write(partitionInfo.getLocalReplicaId().getMountPath(), false);
+      } catch (IOException | ReplicationException e) {
+        logger.warn(
```

Comment:
Minor: exceptions like this should still be logged as error, even if we're not rethrowing.